feat(service-worker): handle 'notificationclick' events

joostme opened the pull request:
## PR Checklist

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
The current version of the service worker does not allow to handle clicks on the notification or be informed which notification or action was clicked.

Issue Number: #20956, #22311


## What is the new behavior?
The new behaviour also listens for the 'notificationclick' event and provides an observable of all messages that were clicked in the `SwPush` service.

```ts
swPush.messagesClicked.pipe(
  ...
);
```

The emitted values will be of the following type:
```ts
{
  action: string,
  notification: NotificationObject
}
```
where `NotificationObject` is the following interface:
```ts
export interface NotificationObject extends NotificationOptions { title: string; }
```

If no custom actions are defined in the notification, action will be `undefined`.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

joostme committed:
feat(service-worker): handle 'notificationclick' events

The previous version did not support the 'notificationclick' event.
Add event handler for the event and provide an observable of
clicked notifications in the SwPush service.

Closes #20956, #22311

joostme committed:
fix(service-worker): add missing api typing

Add missing messagesClicked typing for SwPush service

gkalpak commented on packages/service-worker/test/comm_spec.ts:
```diff
@@ -303,6 +303,26 @@ import {async_fit, async_it} from './async';
+              (msg: {message: string}) => receivedMessages.push(msg.message));
+
+          sendMessage('NOTIFICATION_CLICK', 'this was a click');
+          sendMessage('NOT_OTIFICATION_CLICK', 'this was not a click');
```
This should either be `NOT_NOTIFICATION_CLICK` or `NOT_IFICATION_CLICK` (for extra fun :smiley:).

gkalpak commented on packages/service-worker/worker/src/service-worker.d.ts:
```diff
@@ -83,6 +83,8 @@ interface PushEvent extends ExtendableEvent {
   data: PushMessageData;
 }
 
+interface NotificationClickEvent extends NotificationEvent, ExtendableEvent {}
```
AFAICT, there is no such thing a `NotificationClickEvent`. It is actually the `NotificationEvent` , so maybe we should just append `extends ExtendableEvent` to the `NotificationEvent` interface above.

gkalpak commented on packages/service-worker/worker/testing/scope.ts:
```diff
@@ -333,6 +342,9 @@ class MockPushEvent extends MockExtendableEvent {
     json: () => this._data,
   };
 }
+class MockNotificationClickEvent extends MockExtendableEvent {
```
Nit: `MockNotificationEvent` is probably more accurate.

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -286,6 +292,13 @@ export class Driver implements Debuggable, UpdateSource {
+      type: 'NOTIFICATION_CLICK',
+      data: {action, notification},
+    });
+  }
```
It would probably make sense to close the notification here two.

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -286,6 +292,13 @@ export class Driver implements Debuggable, UpdateSource {
+    await this.broadcast({
+      type: 'NOTIFICATION_CLICK',
+      data: {action, notification},
+    });
```
A couple of questions:

1. Will the client actually have access to the notification object? (That would be cool.)
2. What if there are no clients open?
3. It might be worth supporting opening a new client (potentially passing the action as a query param or something).
    _Not something that needs to happen on this PR though._

joostme replied:
Wow! Thanks for the great feedback @gkalpak . 

Regarding your questions:
1. That's sadly not possible. The notification object is passed to the client window using *postmessage* . This serializes all data and fails when serializing the notification object as it has functions. So a similar approach as used on lines 288-291 should be used to copy only some properties. I tested this locally and it failed when trying to serialize the notification object 😕 I will implement the manual serialization in the coming commits.
2. Currently if no clients are open, nothing will happen. Maybe the service worker should open a new window if a client has no open window and the notification has a `url` property or something.
3. I like the idea and will think about how to solve this. I might open a new PR for this

Again. Thanks for the feedback 😊. I will provide the changes as soon as I find the time this week.

gkalpak reviewed (commented):
I think this is heading in the right direction. Thx, @joostme 👍
I left a few commets/questions.

joostme committed:
refactor(service-worker): update tests based of review feedback

- Rename invalid click event name
- Make test expects more explicit
- Remove NotificationClickEvent in favor of NotificationEvent

joostme committed:
feat(service-worker): close notifications and focus window on click

- Serialize notification object before using postMessage
- Close notification on click
- Focus browser if it is not already focused on click

joostme commented:
@gkalpak I implemented your suggested changes 😊

Furthermore the client is now focused if a notification is clicked and the client allows focusing.
I tested all this locally and it also works with the manual serialization of the notification.

I think opening the browser window or navigating to a specific url should also be possible. Especially if a notification is clicked while no window is active or opened. But that is something for another PR in my opinion.

JonnyBGod commented:
For your inspiration:
```
this.scope.addEventListener('notificationclick', (event) => {
            event.notification.close();
            event.waitUntil(clients.matchAll({
              type: "window"
            }).then(function(clientList) {
              const link =
                event.notification.data &&
                event.notification.data.fcmOptions &&
                event.notification.data.fcmOptions.link;

              for (var i = 0; i < clientList.length; i++) {
                var client = clientList[i];
                if (link && link === client.url) {
                  return client.focus();
                } else if ('focus' in client) {
                  if (link) {
                    client.navigate(link);
                  }
                  return client.focus();
                }
              }
              if (clients.openWindow) {
                return clients.openWindow(link || '/');
              }
            }));
          });
```
this is working for me.

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -950,6 +974,18 @@ export class Driver implements Debuggable, UpdateSource {
+    clients.forEach((client: any) => {
+      if ('focus' in client) {
+        if (!client.focused) {
+          client.focus();
```
According to w3c/ServiceWorker#602 (and my experiments) you can only call one method of `clients.openWindow()` or `client.focus()` (so you can't call `.focus()` on all clients). Why would you do that anyway?

We need to think this through. Let's keep it simple in this PR (and avoid having to introduce a breaking change in the future).
I suggest we only broadcast and let the apps deal with the message.

##
In a future PR, we can discuss other options. Off the top of my head:

- If there are no open clients:
  1. Open a new window at `ServiceWorkerRegistration#scope`.
  2. Open a new window at a URL defined in `Notification#data.url` (or some other property we agree on).
  3. Have another convention to specify URLs per action in `Notification#data.urlPerAction`. Then we could look up the URL to open based on the action's name.

- If there are clients open:
  1. Focus the last focused client. 
  2. Navigate the last focused client to a specific URL (e.g. if specified in `Notification#data.url` or `Notification#data.urlPerAction`).

- Alternatively, we could specify some operations per action according to some predefined convention. E.g.:
    ```ts
    notification: NotificationOptions = {
      actions: [
        {action: 'foo', title: 'Open new tab'},
        {action: 'bar', title: 'Navigate in last focused tab'},
        {action: 'baz', title: 'Focus last focused tab'},
        {action: 'qux', title: 'Focus last focused tab (if any) or open new tab'},
      ],
      data: {
        onActionClick: {
          foo: {operation: 'openWindow', url: 'http://example.com/foo'},
          bar: {operation: 'navigateLastFocused', url: 'http://example.com/bar'},
          baz: {operation: 'focusLastFocused'},
          qux: {operation: 'focusLastFocusedOrOpen', url: 'http://example.com/qux'},
        },
      },
    }
    ```

glebmachine replied:
@gkalpak I hope you keep working on `openWindow`
It's very important to support this feature, otherside you unable to use swPush service in production(

I will be happy to propose to use following strategy:
1. Pre-check if URL provided 
2. If app has opened windows - focus last one, and follow received url
3. If don't - open new window with received url

You don't need to overcomplicate sw with onActionClick. 
My proposed strategy looks generic and has expected behaviour

glebmachine replied:
But, actually, best choice is always open new tab. I have really bad user experience with Basecamp Push Notifications.
Each time when i'm accidentally click to Push Notification it navigate one of my "attached" pages with tasks to new address. So it's very bad.

So, conclusion, if i click to notification, i assume that it will open new window (don't interrupts my current session)

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -21,6 +21,7 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel} from './low_level';
 @Injectable()
 export class SwPush {
   readonly messages: Observable<object>;
+  readonly messagesClicked: Observable<object>;
```
We should have some API docs for this (see #23138).
We should mention which properties the payload will have (`action` and `notification`) and make it clear that the `notification` is not a [Notification][1] object (but more like a [NotificationOptions][2] obejct).

[1]: https://developer.mozilla.org/en-US/docs/Web/API/Notification
[2]: https://notifications.spec.whatwg.org/#dictdef-notificationoptions

gkalpak reviewed (changes requested):
Getting there 👍
Let's keep this PR simple (and safe for future modification) and then discuss more advanced usecases in another issue/PR.

joostme committed:
refactor(service-worker): Code optimizations

- Add missing new line in tests
- Add proper noop method in MockNotificationEvent

joostme committed:
feat(service-worker): Add typing for messagesClicked in SwPush service

- Properly type messagesClicked Observable stream
- Add new NotificationObject interface

joostme committed:
refactor(service-worker): Rework notification click handler

- Add missing image and timestamp properties
- Remove focus from click handler

joostme committed:
fix(service-worker): Add typing to public api guard and fix lint errors

joostme commented:
@gkalpak I updated the pull request with according to your feedback. Again thank you for your great feedback 😊!

- I removed the focus of the window for now to keep things simple in this PR. The click handler now uses the normal `broadcast` method.
- I also properly typed the `messagesClicked` stream in the `SwPsuh` service and added JSDocs.
- `image` and `timestamp` are now added to the Notification object in the `messagesClicked` stream.

The `handleClick` method copies the notification object and creates an `options` object because the `broadcast` method uses `postMessage`. This serializes the data but is unable to serialize the notification object as it contains functions. So I copy the necessary properties into the options object. The same solution is also used in the `handlePush` method above.

joostme committed:
fix(service-worker): Fix public api guard typing

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -33,6 +33,10 @@ const NOTIFICATION_OPTION_NAMES = [
   'actions', 'badge', 'body', 'dir', 'icon', 'lang', 'renotify', 'requireInteraction', 'tag',
   'vibrate', 'data'
 ];
+const NOTIFICATION_CLICK_OPTION_NAMES = [
```
Actually, why have a different array and not re-use `NOTIFICATION_OPTION_NAMES`?
The only difference should be `title`.

We can either include `title` (since it won't matter) or omit it from `NOTIFICATION_OPTION_NAMES` and have a second list which adds `title`:

```ts
const NOTIFICATION_OPTION_NAMES = [...];
const NOTIFICATION_PROPERTY_NAMES = NOTIFICATION_OPTION_NAMES.concat('title');
```

gkalpak reviewed (changes requested):
Thx for making the changes, @joostme 👍
Here's another round of feedback (since you seem to be enjoying it) 😛

joostme commented on packages/service-worker/src/push.ts:
```diff
@@ -21,6 +21,21 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel} from './low_level';
+    action: string;
+    notification: NotificationObject
+  }
+  > ;
```
Yeah the clang formatter formatted it like this 🤔. I know it looks kind of strange

joostme commented on packages/service-worker/src/push.ts:
```diff
@@ -21,6 +21,21 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel} from './low_level';
+   * Note that the `notification` property is __not__ a
+   * [Notification](https://developer.mozilla.org/en-US/docs/Web/API/Notification) but rather a
+   * [NotificationOptions](https://notifications.spec.whatwg.org/#dictdef-notificationoptions)
+   * object that also includes the notification `title`.
```
The title comes from the Notification object. I will add this so it is more clear

joostme commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -286,6 +296,20 @@ export class Driver implements Debuggable, UpdateSource {
+  private async handleClick(notification: Notification, action?: string): Promise<void> {
+    notification.close();
+
+    const desc = notification as any;
```
This will not work as Notification includes a lot more properties than those listed in `NOTIFICATION_OPTION_NAMES` like `onclick`, `onclose` etc.

gkalpak replied:
Which part won't work? And why?
(I know `keyof Notification` is a superset of the allowed values, but it is still more strict than the currently inferred `string[]` type.)

joostme replied:
I now actually got this working 👏. The `NOTIFICATION_OPTION_NAMES` is now of type `(keyof Notification)[]` and the `options` object of type `{-readonly: [key in keyof Notification]?: Notification[key]}`.

The problem I had before were the missing properties in the `NOTIFICATION_OPTION_NAMES` array. But the TS error I got was so misleading I did not catch that.

New version is on its way!

mary-poppins commented:
You can preview 9db790a at https://pr25860-9db790a.ngbuilds.io/.

joostme commented:
I refactored it and did the following changes:

- Rename `messagesClicked` to `notificationClicks`
- Update JSDoc for `notificationClicks` to be more precise on where the `title` comes from
- Add missing properties to the `NOTIFICATION_OPTION_NAMES` array and use this one instead of a new one just for the `handleClick` method
- Use intersection type `NotificationOptions & { title: string }` in `SwPush` service

I was not able to remove the `any` cast in the `handleClick` method even with your super fancy typing for `options`. Typescript is still giving me warnings when not casting it to `any` because of a missing index signature on the `Notification` interface 😐

mary-poppins commented:
You can preview 659dd95 at https://pr25860-659dd95.ngbuilds.io/.

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -339,12 +361,14 @@ export class Driver implements Debuggable, UpdateSource {
 
-    // If initialization fails, the SW needs to enter a safe state, where it declines to respond to
+    // If initialization fails, the SW needs to enter a safe state, where it declines to respond
+    // to
```
Did the formatter do that? Why? AFAICT, nothing has changed on these lines 😕

gkalpak reviewed (changes requested):
Really close now :smiley:

mary-poppins commented:
You can preview 0289564 at https://pr25860-0289564.ngbuilds.io/.

joostme commented:
I made the changes you suggested @gkalpak . 😊👏

- The `NOTIFICATION_OPTION_NAMES` now has the proper `(keyof Notification)[]` typing
- The `options` object now has the `{-readonly: [key in keyof Notification]?: Notification[key]}` typing to be absolutely correct.
- I removed the `NotificationObject` interface in favor of the existing `NotificationOptions` interface
- The filter in `handleClick` now uses `name in notification` instead of `hasOwnProperty` as `hasOwnProperty` actually always returns `false` because the properties only exist in the `Notification` prototype. I tested this locally and it works properly now.
- I reverted the unwanted line-breaks in the comments. My tslint wanted to format those lines 😶

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -286,6 +292,19 @@ export class Driver implements Debuggable, UpdateSource {
+  private async handleClick(notification: Notification, action?: string): Promise<void> {
+    notification.close();
+
+    const options: {-readonly[K in keyof Notification]?: Notification[K]} = {};
```
For the record, `options: {[key: string]: any}` would be enough here. IIRC, the `-readonly` syntax is not supported in TS 2.7.x, so it might not be compatible with the current stable branch, but since this is a feature and will only go into the `@next` branch (i.e. master), this is probably fine (and fancy 😛).

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -543,6 +543,18 @@ const manifestUpdateHash = sha1(JSON.stringify(manifestUpdate));
+    async_it('broadcasts notification click events', async() => {
+      expect(await makeRequest(scope, '/foo.txt')).toEqual('this is foo');
+      await driver.initialized;
+      await scope.handleClick({title: 'This is a test', body: 'Test body'}, 'button');
```
Can you also add a call to `handleClick()` without an action (to verify that the SW can handle that)?

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -286,6 +292,19 @@ export class Driver implements Debuggable, UpdateSource {
+    notification.close();
+
+    const options: {-readonly[K in keyof Notification]?: Notification[K]} = {};
+    NOTIFICATION_OPTION_NAMES.filter(name => name in notification)
```
It might be worth adding a comment here, explaining that notification properties might be on the prototype and not the notification object directly.
Or ensure that tests would fail (i.e. set the properties on the prototype of `MockNotificationEvent#notification`).

gkalpak reviewed (changes requested):
Almost there :smiley: Keep it up 😛

mary-poppins commented:
You can preview ca42261 at https://pr25860-ca42261.ngbuilds.io/.

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -21,6 +21,24 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel} from './low_level';
+   * If no action was used the action property will be an empty string `''`.
+   *
+   * Note that the `notification` property is **not** a
+   * [Notification][Mozilla Notification]  object but rather
```
- Can't the link go to the line above?
- Double space before "object".

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -21,6 +21,24 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel} from './low_level';
+   * object that also includes the `title` of the
+   * [Notification][Mozilla Notification] object.
+   *
+   * [Mozilla Notification]: https://developer.mozilla.org/en-US/docs/Web/API/Notification
```
I didn't know this syntax works in API docs comments :tada:

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -21,6 +21,24 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel} from './low_level';
+   * [Notification][Mozilla Notification]  object but rather
+   * a [NotificationOptions](https://notifications.spec.whatwg.org/#dictdef-notificationoptions)
+   * object that also includes the `title` of the
+   * [Notification][Mozilla Notification] object.
```
Can you please format these lines to have as few of them as possible?

gkalpak reviewed (approved):
A couple of comment formatting. Other than that LGTM 👌

joostme committed:
refactor(service-worker): Format comments and add additional test

- Format JSDoc for notificationClicks
- Add comment on why handleClick does not use hasOwnProperty
- Add additional test that uses handleClick without action

mary-poppins commented:
You can preview 67e4b8c at https://pr25860-67e4b8c.ngbuilds.io/.

joostme commented:
@gkalpak I changed the formatting of the JSDoc. Not everything could be optimized because of the clang formatter. But it's a bit cleaner now ☺ 
Thank you for taking so much time and providing great feedback 👍🏻

gkalpak reviewed (approved):
LGTM :100:
Thank you for working on this and following through :heart:

Looking forward to the next PR (https://github.com/angular/angular/pull/25860#discussion_r219509018 maybe?) :wink:

IgorMinar commented on packages/service-worker/src/push.ts:
```diff
@@ -21,6 +21,22 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel} from './low_level';
   readonly messages: Observable<object>;
+  /**
+   * Emits the payloads of the received push notification messages as well as the action the user
+   * interacted with. If no action was used the action property will be an empty string `''`.
```
why empty string and not undefined as is the case in the DOM api?

gkalpak replied:
MDN ~~says~~ _used to say(*)_ that [`action` will be undefined][1], but [according to the spec][2] (which Chrome's implementation matches atm) it will be an empty string:

> If `notification` is a persistent notification, then:
>   - **Let `action` be the empty string.**
>   - If one of `notification`’s actions was activated by the user, then set `action` to that action’s name.
>   - ...

<sub>*: I updated MDN (😱)</sub>

[1]: https://developer.mozilla.org/en-US/docs/Web/API/NotificationEvent/action
[2]: https://notifications.spec.whatwg.org/#activating-a-notification

IgorMinar replied:
Awesome! 🎉

IgorMinar reviewed (commented):
The rest looks great - can you please clarify the question about `undefined` below?

We are already in RC for 7.0. So we'll merge this once master is open for 7.1 changes in a week or so.

mary-poppins commented:
You can preview a200d92 at https://pr25860-a200d92.ngbuilds.io/.

mary-poppins commented:
You can preview a5f3587 at https://pr25860-a5f3587.ngbuilds.io/.

amjmhs commented on packages/service-worker/src/push.ts:
```diff
@@ -45,9 +62,13 @@ export class SwPush {
+    this.notificationClicks =
+        this.sw.eventsOfType('NOTIFICATION_CLICK').pipe(map((message: any) => message.data));
 
     this.messages = this.sw.eventsOfType<PushEvent>('PUSH').pipe(map(message => message.data));
```
Isn't line 69 and line 73 the same ?

gkalpak commented:
Yes, that's righ, @lorado. For the sake of keeping this PR simple, it will have no effect if no app instances are running.
Once this PR is merged, we can think about and work on more advanced features. (E.g. see https://github.com/angular/angular/pull/25860#discussion_r219509018.)

gkalpak commented:
@IgorMinar, is this unblocked now?

IgorMinar reviewed (approved):
yup. unblocked.

IgorMinar commented:
I've rebased this PR. hopefully the CI will go green. If not, we'll need to investigate.

mary-poppins commented:
You can preview cd8d801 at https://pr25860-cd8d801.ngbuilds.io/.

gkalpak commented:
:tada: Nice job, @joostme :100:

(BTW, I opened #26907 for discussing possible future enhancements. Feel free to chime in :wink:)
